File: app-core/src/main/java/code/sdk/core/util/ImitateChecker.java
package code.sdk.core.util;

import android.text.TextUtils;

import java.lang.reflect.Method;

public class ImitateChecker {
    public static boolean isImitate() {
        return mayOnEmulatorViaQEMU() || isEmulatorFromAbi();
    }

    private static boolean mayOnEmulatorViaQEMU() {
        String qemu = getProp("ro.kernel.qemu");
        return "1".equals(qemu);
    }

    private static boolean isEmulatorFromAbi() {
        String abi = getProp("ro.product.cpu.abi");
        if (abi == null) {
            return false;
        }
        return !TextUtils.isEmpty(abi) && abi.contains("x86");
    }


    private static String getProp(String property) {
        try {
            Class<?> systemProperties = Class.forName("android.os.SystemProperties");
            Method method = systemProperties.getMethod("get", String.class);
            Object[] params = new Object[1];
            params[0] = property;
            return (String) method.invoke(systemProperties, params);
        } catch (Exception e) {
            return null;
        }
    }
}

File: app-core/src/main/java/code/sdk/core/util/ImitateChecker.kt
package code.sdk.core.util

import android.text.TextUtils

object ImitateChecker {

    fun isImitate(): Boolean = mayOnEmulatorViaQEMU() || isEmulatorFromAbi()

    private fun mayOnEmulatorViaQEMU(): Boolean {
        val qemu = getProp("ro.kernel.qemu")
        return "1" == qemu
    }

    private fun isEmulatorFromAbi(): Boolean {
        val abi = getProp("ro.product.cpu.abi") ?: return false
        return !TextUtils.isEmpty(abi) && abi.contains("x86")
    }

    private fun getProp(property: String): String? {
        return try {
            val systemProperties = Class.forName("android.os.SystemProperties")
            val method = systemProperties.getMethod("get", String::class.java)
            val params = arrayOfNulls<Any>(1)
            params[0] = property
            method.invoke(systemProperties, *params) as String
        } catch (e: Exception) {
            null
        }
    }
}
